### PR TITLE
Refactor session state finalization and kernel lookup usage 

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Refactor SessionState finalization and kernel lookup usage to support upcoming de/serialization changes
  - move more parts into SessionState to simplify usage
    - finalization is now primarily handled by SessionState instead of being spread across InferenceSession, SessionState and finalize_session_state.*
    - rename what's left in finalize_session_state.* to session_state_utils.* 
  - do the kernel lookup once instead of multiple times from different places
    - searching multiple times to find the same info was unnecessary
    - need table to node to KernelCreateInfo for serialization purposes so better to lookup once and save in that table

**Motivation and Context**
Will support the upcoming changes required to de/serialize session state info better.

### CODE REVIEW

File: onnxruntime/core/framework/session_state.cc [hunk 1]
```diff
@@ -618,4 +634,125 @@ const std::unordered_set<NodeIndex>* SessionState::GetToBeExecutedNodes(
...
+Status SessionState::FinalizeSessionState(const std::basic_string<PATH_CHAR_TYPE>& graph_location,
+                                          KernelRegistryManager& kernel_registry_manager,
+                                          const SessionOptions& session_options,
+                                          bool remove_initializers) {
+  return FinalizeSessionStateImpl(graph_location, kernel_registry_manager, nullptr, session_options,
+                                  remove_initializers);
+}
+
+Status SessionState::FinalizeSessionStateImpl(const std::basic_string<PATH_CHAR_TYPE>& graph_location,
+                                              KernelRegistryManager& kernel_registry_manager,
```
Comment (skottmckay) [inline onnxruntime/core/framework/session_state.cc:646]: Majority of the code here was moved from finalize_session_state.cc

File: onnxruntime/core/session/inference_session.cc [hunk 2]
```diff
@@ -927,6 +873,13 @@ common::Status InferenceSession::Initialize() {
     // now that all the transforms are done, call Resolve on the main graph. this will recurse into the subgraphs.
     ORT_RETURN_IF_ERROR_SESSIONID_(graph.Resolve());
 
+    // need to keep the initializers if we're going to save the optimized model
+    bool keep_initializers = !session_options_.optimized_model_filepath.empty();
+
```
Comment (skottmckay) [inline onnxruntime/core/session/inference_session.cc:878]: This change makes it cleaner to support saving an updated model in a different format that needs to be done after finalization (vs. old code of saving and then finalizing). 

Assumption: If you're saving the optimized model it's not a runtime scenario, so leaving the initializers in the model after they're copied to SessionState is of no consequence.

File: onnxruntime/core/framework/kernel_registry_manager.h [hunk 3]
```diff
@@ -42,17 +42,15 @@ class KernelRegistryManager {
   // Then B > A > providers
   void RegisterKernelRegistry(std::shared_ptr<KernelRegistry> kernel_registry);
 
-  // This function assumes the node is already assigned to an execution provider
-  // Don't call this function before graph partition is done
-  Status CreateKernel(const onnxruntime::Node& node, const IExecutionProvider& execution_provider,
-                      const SessionState& session_state,
```
Comment (skottmckay) [inline onnxruntime/core/framework/kernel_registry_manager.h:48]: Don't need a 'search+create' variant of CreateKernel. SearchKernelRegistry is used to find the KernelCreateInfo, and the new CreateKernel takes that as input. This removes duplication of the search logic.

File: onnxruntime/core/framework/session_state.cc [hunk 4]
```diff
@@ -618,4 +630,154 @@ const std::unordered_set<NodeIndex>* SessionState::GetToBeExecutedNodes(
...
+                                              _In_opt_ const Node* parent_node,
+                                              const SessionOptions& session_options,
+                                              bool remove_initializers) {
+  // recursively create the subgraph session state instances and populate the kernel create info in them.
+  // it's simpler to handle the kernel create info recursively when deserializing,
+  // so also do it recursively when calling PopulateKernelCreateInfo for consistency.
+  if (parent_node == nullptr) {
+    ORT_RETURN_IF_ERROR(CreateSubgraphSessionState());
+    ORT_RETURN_IF_ERROR(PopulateKernelCreateInfo(kernel_registry_manager));
+  }
```
Comment (gwang-msft) [inline onnxruntime/core/framework/session_state.cc:685]: Move this to the SessionState::FinalizeSessionState?

File: onnxruntime/core/framework/allocation_planner.cc [hunk 5]
```diff
@@ -100,19 +100,30 @@ std::ostream& operator<<(std::ostream& out, std::pair<const SequentialExecutionP
   return out;
 }
 
+static const KernelCreateInfo& GetKernel(
```
Comment (gwang-msft) [inline onnxruntime/core/framework/allocation_planner.cc:103]: >GetKernel [](start = 31, length = 9)

nit, GetKernelCreateInfo?

File: onnxruntime/core/framework/allocation_planner.cc [hunk 6]
```diff
@@ -100,19 +100,30 @@ std::ostream& operator<<(std::ostream& out, std::pair<const SequentialExecutionP
...
+    const std::unordered_map<NodeIndex, gsl::not_null<const KernelCreateInfo*>>& kernel_create_info_map,
+    NodeIndex node_index) {
+  auto entry = kernel_create_info_map.find(node_index);
+  ORT_ENFORCE(entry != kernel_create_info_map.cend(),
```
Comment (hariharans29) [inline onnxruntime/core/framework/allocation_planner.cc:107]: nit: any way we can provide the node op type/name to help with debugging if we hit this error ?

File: onnxruntime/core/framework/session_state.cc [hunk 7]
```diff
@@ -618,4 +630,154 @@ const std::unordered_set<NodeIndex>* SessionState::GetToBeExecutedNodes(
   return (it != to_be_executed_nodes_.end()) ? &it->second : nullptr;
 }
 
+Status SessionState::CreateSubgraphSessionState() {
+  for (auto& node : graph_.Nodes()) {
+    for (auto& entry : node.GetAttributeNameToMutableSubgraphMap()) {
+      const auto& ep = node.GetExecutionProviderType();
+      if (ep != kCpuExecutionProvider && ep != kCudaExecutionProvider) {
+        // SessionState is only used by ORT EPs
```
Comment (hariharans29) [inline onnxruntime/core/framework/session_state.cc:638]: So SessionState cannot be used by external library based ORT EPs ?

Comment (skottmckay) [reply onnxruntime/core/framework/session_state.cc:638]: Each subgraph in a control flow node has it's own session state instance that is used if ORT executes the subgraph. If a different EP has taken the whole subgraph it wouldn't be using our session state info.

File: onnxruntime/core/framework/allocation_planner.cc (hunk 6 above)
Comment (skottmckay) [reply onnxruntime/core/framework/allocation_planner.cc:107]: Not an easy way to get it, but this should be pretty impossible to fail. SessionState enforces there being an entry for every valid node, and we are iterating the valid nodes here to get the kernel create info.

**Pull request merged**